lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemStats.java

```diff
@@ -0,0 +1,222 @@
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+import static software.amazon.awssdk.core.internal.metrics.SdkErrorType.SERVER_ERROR;
+import static software.amazon.awssdk.core.internal.metrics.SdkErrorType.THROTTLING;
+import static software.amazon.awssdk.core.metrics.CoreMetric.API_CALL_DURATION;
+import static software.amazon.awssdk.core.metrics.CoreMetric.API_CALL_SUCCESSFUL;
+import static software.amazon.awssdk.core.metrics.CoreMetric.ERROR_TYPE;
+import static software.amazon.awssdk.core.metrics.CoreMetric.OPERATION_NAME;
+import static software.amazon.awssdk.core.metrics.CoreMetric.RETRY_COUNT;
+import static software.amazon.awssdk.core.metrics.CoreMetric.SERVICE_ID;
+
+public class S3FileSystemStats
+{
+    private final AwsSdkV2ApiCallStats total = new AwsSdkV2ApiCallStats();
+
+    private final AwsSdkV2ApiCallStats getObject = new AwsSdkV2ApiCallStats();
+    private final AwsSdkV2ApiCallStats listObjectsV2Paginator = new AwsSdkV2ApiCallStats();
+    private final AwsSdkV2ApiCallStats putObject = new AwsSdkV2ApiCallStats();
+    private final AwsSdkV2ApiCallStats createMultipartUpload = new AwsSdkV2ApiCallStats();
+    private final AwsSdkV2ApiCallStats completeMultipartUpload = new AwsSdkV2ApiCallStats();
+    private final AwsSdkV2ApiCallStats abortMultipartUpload = new AwsSdkV2ApiCallStats();
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getTotal()
+    {
+        return total;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getGetObject()
+    {
+        return getObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getListObjectsV2Paginator()
+    {
+        return listObjectsV2Paginator;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getPutObject()
+    {
+        return putObject;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCreateMultipartUpload()
+    {
+        return createMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getCompleteMultipartUpload()
+    {
+        return completeMultipartUpload;
+    }
+
+    @Managed
+    @Nested
+    public AwsSdkV2ApiCallStats getAbortMultipartUpload()
+    {
+        return abortMultipartUpload;
+    }
+
+    public MetricPublisher newMetricPublisher()
+    {
+        return new JmxMetricPublisher(this);
+    }
+
+    public final class JmxMetricPublisher
+            implements MetricPublisher
+    {
+        private static final Logger log = Logger.get(JmxMetricPublisher.class);
+
+        private final S3FileSystemStats stats;
+
+        public JmxMetricPublisher(S3FileSystemStats stats)
+        {
+            this.stats = requireNonNull(stats, "stats is null");
+        }
+
+        private static final Set<SdkMetric<?>> ALLOWED_METRICS = Set.of(API_CALL_SUCCESSFUL, RETRY_COUNT, API_CALL_DURATION, ERROR_TYPE);
+
+        private static final Set<String> ALLOWED_SERVICE_IDS = Set.of("S3");
```

Comment:
Let's inline this since there is only one value

Reply:
Updated in the revision.